File: test/ModularPipelines.UnitTests/Models/CommandLineTests.cs
using ModularPipelines.Models;

namespace ModularPipelines.UnitTests.Models;

public class CommandLineTests
{
    [Test]
    public async Task CommandLine_StoresToolAndArguments()
    {
        var commandLine = new CommandLine("git", ["add", "--all"]);

        using (Assert.Multiple())
        {
            await Assert.That(commandLine.Tool).IsEqualTo("git");
            await Assert.That(commandLine.Arguments).IsEquivalentTo(new[] { "add", "--all" });
        }
    }

    [Test]
    public async Task CommandLine_ToString_FormatsCorrectly()
    {
        var commandLine = new CommandLine("git", ["commit", "-m", "test message"]);

        await Assert.That(commandLine.ToString()).IsEqualTo("git commit -m test message");
    }

    [Test]
    public async Task CommandLine_EmptyArguments_ToStringShowsToolOnly()
    {
        var commandLine = new CommandLine("git", []);

        await Assert.That(commandLine.ToString()).IsEqualTo("git");
    }

    [Test]
    public async Task CommandLine_ArgumentsAreImmutable()
    {
        var args = new List<string> { "add" };
        var commandLine = new CommandLine("git", args);
        args.Add("--all");

        await Assert.That(commandLine.Arguments).HasCount().EqualTo(1);
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Models/CommandLineTests.fs
namespace ModularPipelines.UnitTests.FSharp.Models

open System.Collections.Generic
open ModularPipelines.Models
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type CommandLineTests() =
    [<Test>]
    member _.CommandLine_StoresToolAndArguments() = async {
        let commandLine = CommandLine("git", ["add"; "--all"])
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(commandLine.Tool), "git"))
        do! check(Assert.That((commandLine.Arguments |> Seq.toArray) = [| "add"; "--all" |]).IsTrue())
    }

    [<Test>]
    member _.CommandLine_ToString_FormatsCorrectly() = async {
        let commandLine = CommandLine("git", ["commit"; "-m"; "test message"])
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(commandLine.ToString()), "git commit -m test message"))
    }

    [<Test>]
    member _.CommandLine_EmptyArguments_ToStringShowsToolOnly() = async {
        let commandLine = CommandLine("git", [])
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(commandLine.ToString()), "git"))
    }

    [<Test>]
    member _.CommandLine_ArgumentsAreImmutable() = async {
        let args = List<string>(["add"])
        let commandLine = CommandLine("git", args)
        args.Add("--all")
        do! check(Assert.That(commandLine.Arguments |> Seq.length = 1).IsTrue())
    }
